cis194/week4/zhansongl/Wholemeal.hs

```diff
@@ -0,0 +1,71 @@
+module Wholemeal where
+
+import Data.Bool (bool)
+import qualified Data.Array as A
+
+fun1 :: [Integer] -> Integer
+fun1 [] = 1
+fun1 (x:xs)
+  | even x    = (x - 2) * fun1 xs
+  | otherwise = fun1 xs
+
+fun1' :: [Integer] -> Integer
+fun1' = foldr (*) 1 . map (flip (-) 2) . filter even
+
+fun2 :: Integer -> Integer
+fun2 1 = 0
+fun2 n | even n    = n + fun2 (n `div` 2)
+       | otherwise = fun2 (3 * n + 1)
+
+fun2' :: Integer -> Integer
+fun2' = sum . filter even . takeWhile (/= 1) . iterate next
+  where next n = bool (3 * n + 1) (n `div` 2) $ even n
+
+
+data Tree a = Leaf
+            | Node Integer (Tree a) a (Tree a)
+  deriving (Show, Eq)
+
+getHeight :: Tree a -> Integer
+getHeight Leaf = (-1)
+getHeight (Node h _ _ _) = h
+
+insert :: a -> Tree a -> Tree a
+insert x Leaf = Node 0 Leaf x Leaf
+insert x (Node h left y right)
+  | getHeight left < getHeight right =
+      let h' = 1 + max (getHeight left') (getHeight right)
+          left' = insert x left
+      in  Node h' left' y right
+  | otherwise =
+      let h' = 1 + max (getHeight left) (getHeight right')
+          right' = insert x right
+      in  Node h' left y right'
```

Comment:
For the case of `left` and `right` being the same height, is it correct to always insert on the `right`?

Reply:
it should be, because inserting a single element can increase the height of a tree by at most one, I think

Reply:
When `left` and `right` being the same height, is it possible that inserting to the `left` does not increase the height and inserting to the `right` does?

Reply:
yes, in which case it will still be balanced because the height difference is 1

Reply:
So following my previous comment, when it’s inserting to the right the new node underneath will be 2 levels deeper than the shorter leg under left. No?

Reply:
I think I understand what you mean now, will fix

Reply:
hmm, actually, I still think it's correct because if you look at the testing code [here](https://github.com/fp-works/2019-winter-Haskell-school/pull/70/files#diff-b43f2ea241398742a3d1444be175cc08R16) the resulting tree should be balanced given an arbitrary list which should be enough proof that it is correct.

Reply:
in the second case the height of the left tree should be equal or greater than the right, which means inserting to the right will increase the height difference by at most one, and will still result in a balanced tree